### PR TITLE
Fix portability issue that prevents compilation on some systems

### PR DESCRIPTION
**Pull request by ghost**
On February 4th, @JeffBezanson accepted a commit from @carnaval that introduced portability issues.
https://github.com/JuliaLang/julia/commit/a09fd8dc0e396362744149977ccfe0b3b04937fa
Especifically it was the use of RTLD_DL_LINKMAP, which is a GNU extension not present on systems that don't use glibc.
Please, I urge Julia to pay more attention to portability and remember that Linux != GNU (glibc in this case).
There are other C standard libraries out there, such as musl, μClibc, dietlib, bionic...
When we finally got Julia to at least compile by fixing old portability issues, new ones were introduced.
Could we get an official announcement on the mailing list or code guidelines, warning developers not to submit unportable pull requests? To always check if the function they are using is portable or GNU-specific? Maybe even a musl travis check?

I hope we can keep Julia portable.
Thanks for everything.

### CODE REVIEW

**Comment by yuyichao** [comment]
I believe you missed the one below. You'll probably get a variable undefined error otherwise.

**Comment by yuyichao** [comment]
> Could we get an official announcement on the mailing list or code guidelines, warning developers not to submit unportable pull requests?

It is very unlikely that we'll remove those "unportable" functions and we are certainly going to add much more since the portable ones is extremely limited for the level of control we need.

Making sure it build on non-glibc linux is fine but just keep in mind that unless someone with access to those systems is willing to find functionally equivalent to these unportable functions there's almost certainly going to be missing/sub-optimum functions on those setups.

**Comment by StefanKarpinski** [comment]
Policy is unlikely to work effectively – having CI setup for other libcs is the only thing that will work. The JVM-style least common denominator policy is not a good one here – that leads to very limited systems. Instead, we should figure out how to implement various functionality on different systems (like we do across Linux/OS X/Windows) and have as graceful fallbacks as we can on systems that can't.

File: src/debuginfo.cpp [hunk 1]
```diff
@@ -853,12 +853,14 @@ bool jl_dylib_DI_for_fptr(size_t pointer, const llvm::object::ObjectFile **obj,
...
     uint64_t fbase;
-#ifdef __GLIBC__
+#ifdef _OS_LINUX_
     struct link_map *extra_info;
+#ifdef __GLIBC__
     dladdr_success = dladdr1((void*)pointer, &dlinfo, (void**)&extra_info, RTLD_DL_LINKMAP) != 0;
 #else
     dladdr_success = dladdr((void*)pointer, &dlinfo) != 0;
 #endif
+#endif
```
Comment (yuyichao) [inline src/debuginfo.cpp]: This completely changes the meaning AFAIK.

**Comment by albap** [comment]
@yuyichao Could you please propose a fix? I'm not sure how to deal with this.

@StefanKarpinski Are glibc-specific functions so useful that it would be troublesome to start using portable ones instead? I think it would be harder to keep fallbacks for many systems and #ifdefs everywhere than to use standard, portable functions. But either way, I think both methods could work. A musl Travis CI is the best place to start, I think.

**Comment by yuyichao** [comment]
> Could you please propose a fix?

I believe you just need to replace the related (this one and the one below) `_OS_LINUX_` with `__GLIBC__`.

> Are glibc-specific functions so useful that it would be troublesome to start using portable ones instead

Absolutely. The standard is also surprisingly fuzzy in many cases on important details..

**Comment by albap** [comment]
Correct PR here https://github.com/JuliaLang/julia/pull/16797

**Pull request closed without merging**